app/models/user.rb

```diff
@@ -41,12 +41,15 @@ class User < ApplicationRecord
 
   ACTIVE_DURATION = 14.days
 
-  devise :registerable, :recoverable,
-         :rememberable, :trackable, :validatable, :confirmable,
-         :two_factor_authenticatable, :two_factor_backupable,
-         otp_secret_encryption_key: ENV['OTP_SECRET'],
+  devise :two_factor_authenticatable,
```

Comment:
Can you do this? Split devise definitions? That's not obvious to me. What if the last one takes precedence? Why is this change needed anyway?

Reply:
Sorry, should have added a comment for that portion. Yes, that works. In fact it was necessary to get devise to send the notification immediately after the change was requested (and before it's confirmed). Perhaps that's the weird bug you're recalling? I remember checking the commit history and seeing that `config.send_email_changed_notification` was removed when `devise-two-factor` was added - likely because of this. IIRC it's a side-effect of not using `DatabaseAuthenticatable` directly, which is where the notification is implemented, and splitting the definition resolves it. (I don't seem to recall the exact root cause. 😞)

You can check [GitLab](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/models/user.rb) for a similar pattern.

Reply:
Alright, thanks for clearing it up!